### PR TITLE
C++: Suppress ExprHasNoEffect on template code

### LINKED ISSUE #2298: LGTM.com - false positive
**Description of the false positive**

for (Integer i = 0; i < 4; ++i)
     a.row(i) << sPoints[i].x(), sPoints[i].y(), sPoints[i].z(), 1.0;

LGTM gets confused with operator overloading. Here I am using math library Eigen to assign a row to a given matrix.

https://lgtm.com/projects/g/bjaraujo/ENigMA/snapshot/da9570597b4efec3a581d6b37de8a0e64c919319/files/trunk/src/geometry/GeoSphere_Imp.hpp?sort=name&dir=ASC&mode=heatmap#x6c9591050cc7612b:1

### PR DESCRIPTION
**Pull request by jbj**
Fixes #2298 (see https://github.slack.com/archives/CP0MHKV6E/p1574086207203400 for a diagnosis).

Supersedes #2384.

### CODE REVIEW

File: cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql [hunk 1]
```diff
@@ -84,6 +84,7 @@ where
   not peivc.getEnclosingFunction().isDefaulted() and
   not exists(Macro m | peivc = m.getAnInvocation().getAnExpandedElement()) and
   not peivc.isFromTemplateInstantiation(_) and
+  not peivc.isFromUninstantiatedTemplate(_) and
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql]: I'm a bit surprised we didn't already have this exception, given the number of other exceptions!

File: cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql [hunk 2]
```diff
@@ -84,6 +84,7 @@ where
   not peivc.getEnclosingFunction().isDefaulted() and
   not exists(Macro m | peivc = m.getAnInvocation().getAnExpandedElement()) and
   not peivc.isFromTemplateInstantiation(_) and
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql]: Is this line still contributing positively?

Comment (rdmarsh2) [reply cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql]: Yes - it handles cases where a call in a void context has a pure target in some instantiations and a target with side effects in other instantiations. Ideally we'd handle that with a forall, but I don't think we have any way of matching the corresponding elements in different instantiations.

Comment (jbj) [reply cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql]: The unusual thing here is that we're excluding both uninstantiated templates _and_ instantiated templates. That means we're excluding all template code. We usually exclude either one or the other, but in this case there has been false positives in both types of code.

We could try something elaborate like counting the number of potential alerts in the template and its instantiations and reporting an alert only if those counts match, but I don't think this query is important enough for us to spend time on that.

Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql]: OK, sounds like this is a reasonable solution then.

**Pull request merged**